python/cudf/cudf/pandas/_wrappers/pandas.py

```diff
@@ -1734,6 +1736,86 @@ def _unpickle_obj(pickled_args):
     return obj
 
 
+# Save the original __init__ methods
+_original_Series_init = cudf.Series.__init__
+_original_DataFrame_init = cudf.DataFrame.__init__
+_original_Index_init = cudf.Index.__init__
+
+
+def wrap_init(original_init):
+    @functools.wraps(original_init)
+    def wrapped_init(self, data=None, *args, **kwargs):
+        if is_proxy_object(data):
+            data = data.as_gpu_object()
+            if (
+                isinstance(data, type(self))
+                and len(args) == 0
+                and len(kwargs) == 0
+            ):
+                # This short-circuits the constructor to avoid
+                # unnecessary work when the data is already a
+                # proxy object of the same type.
+                # It is a common case in `cuml` and `xgboost`.
+                # For perf impact see:
+                # https://github.com/rapidsai/cudf/pull/17878/files#r1936469215
+                self.__dict__.update(data.__dict__)
+                return
+        original_init(self, data, *args, **kwargs)
+
+    return wrapped_init
+
+
+@functools.wraps(_original_DataFrame_init)
+def DataFrame_init_(self, data, index=None, columns=None, *args, **kwargs):
+    data_is_proxy = is_proxy_object(data)
+
+    if data_is_proxy:
+        data = data.as_gpu_object()
+    elif isinstance(data, list) and len(data) > 0 and is_proxy_object(data[0]):
+        data = [d.as_gpu_object() for d in data]
+    elif (
+        isinstance(data, tuple) and len(data) > 0 and is_proxy_object(data[0])
+    ):
+        data = tuple(d.as_gpu_object() for d in data)
+    elif (
+        isinstance(data, dict)
+        and len(data) > 0
+        and is_proxy_object(next(iter(data.values())))
+    ):
+        data = {k: v.as_gpu_object() for k, v in data.items()}
+
+    if is_proxy_object(index):
+        index = index.as_gpu_object()
+    if is_proxy_object(columns):
+        columns = columns.as_cpu_object()
```

Comment:
Really `as_cpu`?

Reply:
Our `DataFrame.columns` is a `pd.Index` always no?